contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol

```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.16;
+
+import "forge-std/Script.sol";
+
+import "@arbitrum/nitro-contracts/src/precompiles/ArbOwner.sol";
+
+/// @notice should be included in an operation batch in the L1 timelock along with SetWasmModuleRootAction
+contract UpgradeArbOSVersionAtTimestampAction {
+    uint64 public immutable newArbOSVersion;
+    uint64 public immutable upgradeTimestamp;
+
+    constructor(uint64 _newArbOSVersion, uint64 _upgradeTimestamp) {
+        newArbOSVersion = _newArbOSVersion;
+        upgradeTimestamp = _upgradeTimestamp;
+    }
+
+    function perform() external {
+        ArbOwner arbOwner = ArbOwner(0x0000000000000000000000000000000000000070);
+        arbOwner.scheduleArbOSUpgrade({
+            newVersion: newArbOSVersion,
+            timestamp: upgradeTimestamp
+        });
+    }
+}
+
+contract EspressoArbOSUpgrade is UpgradeArbOSVersionAtTimestampAction, Script   {
+    constructor()
+        UpgradeArbOSVersionAtTimestampAction(
+            35,
```

Comment:
when is this called? curious about this hardcoded arbOsVersion

Reply:
This will be called a decent amount of time before the network upgrade is supposed to take place. This will schedule a particular function inside the nodes of an orbit network to run at the given Unix timestamp.

The hardcoded 35 comes from a discussion I had with Mathis when we were first pursuing designating this as an ArbOS upgrade.

Basically, the way ArbOS upgrades are performed is via a switch statement that has some pre-defined gaps in version numbers left open for orbit chains. We chose 35 as it leaves some room for previous upgrades, as well as room for future upgrades in the most recent gap.

Reply:
ahh interesting - let's add that in the comments then, and maybe leave instructions in the readme if it ever has to change